logical/plugin/storage.go

```diff
@@ -0,0 +1,119 @@
+package plugin
+
+import (
+	"net/rpc"
+
+	"github.com/hashicorp/go-plugin"
+	"github.com/hashicorp/vault/logical"
+)
+
+// StorageClient is an implementation of logical.Storage that communicates
+// over RPC.
+type StorageClient struct {
+	client *rpc.Client
+}
+
+func (s *StorageClient) List(prefix string) ([]string, error) {
+	var reply StorageListReply
+	err := s.client.Call("Plugin.List", prefix, &reply)
+	if err != nil {
+		return reply.Keys, err
+	}
+	if reply.Error != nil {
+		return reply.Keys, reply.Error
+	}
+	return reply.Keys, nil
+}
+
+func (s *StorageClient) Get(key string) (*logical.StorageEntry, error) {
+	var reply StorageGetReply
+	err := s.client.Call("Plugin.Get", key, &reply)
+	if err != nil {
+		return nil, err
+	}
+	if reply.Error != nil {
+		return nil, reply.Error
+	}
+	return reply.StorageEntry, nil
+}
+
+func (s *StorageClient) Put(entry *logical.StorageEntry) error {
+	var reply StoragePutReply
+	err := s.client.Call("Plugin.Put", entry, &reply)
+	if err != nil {
+		return err
+	}
+	if reply.Error != nil {
+		return reply.Error
+	}
+	return nil
+}
+
+func (s *StorageClient) Delete(key string) error {
+	var reply StorageDeleteReply
+	err := s.client.Call("Plugin.Delete", key, &reply)
+	if err != nil {
+		return err
+	}
+	if reply.Error != nil {
+		return reply.Error
+	}
+	return nil
+}
+
+// StorageServer is a net/rpc compatible structure for serving
+type StorageServer struct {
+	impl logical.Storage
+}
+
+func (s *StorageServer) List(prefix string, reply *StorageListReply) error {
+	keys, err := s.impl.List(prefix)
+	*reply = StorageListReply{
+		Keys:  keys,
+		Error: plugin.NewBasicError(err),
```

Comment:
`NewBasicError(err)` returns nil if `err` is nil, so `reply.Error` should be nil in that case.